quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -0,0 +1,1348 @@
+            let list_shard_response = ctx
+                .protect_future(self.metastore.list_shards(list_shards_request))
+                .await?;
+
+            self.shard_table
+                .table_entries
+                .reserve(list_shard_response.subresponses.len());
+
+            for list_shards_subresponse in list_shard_response.subresponses {
+                num_shards += list_shards_subresponse.shards.len();
+
+                let key = (
+                    list_shards_subresponse.index_uid.into(),
+                    list_shards_subresponse.source_id,
+                );
+                let shard_entries: HashSet<ShardEntry> = list_shards_subresponse
+                    .shards
+                    .into_iter()
+                    .map(ShardEntry)
+                    .collect();
+                let table_entry = ShardTableEntry {
+                    shard_entries,
+                    next_shard_id: list_shards_subresponse.next_shard_id,
+                };
+                self.shard_table.table_entries.insert(key, table_entry);
+            }
+        }
+        info!(
+            "synced internal state with metastore in {} seconds ({} indexes, {} sources, {} \
+             shards)",
+            now.elapsed().as_secs(),
+            num_indexes,
+            num_sources,
+            num_shards,
+        );
+        Ok(())
+    }
+
+    /// Pings an ingester to determine whether it is available for hosting a shard. If a follower ID
+    /// is provided, the leader candidate is in charge of pinging the follower candidate as
+    /// well.
+    async fn ping_leader(
+        &mut self,
+        ctx: &ActorContext<Self>,
+        leader_id: &NodeId,
+        follower_id_opt: Option<&NodeId>,
+    ) -> Result<(), PingLeaderError> {
+        let mut leader_ingester = self
+            .ingester_pool
+            .get(leader_id)
+            .await
+            .ok_or(PingLeaderError::LeaderUnavailable)?;
+
+        let ping_request = PingRequest {
+            leader_id: leader_id.clone().into(),
+            follower_id: follower_id_opt
+                .cloned()
+                .map(|follower_id| follower_id.into()),
+        };
+        ctx.protect_future(timeout(
+            PING_LEADER_TIMEOUT,
+            leader_ingester.ping(ping_request),
+        ))
+        .await
+        .map_err(|_| PingLeaderError::LeaderUnavailable)? // The leader timed out.
+        .map_err(|error| {
+            if let Some(follower_id) = follower_id_opt {
+                if matches!(error, IngestV2Error::IngesterUnavailable { ingester_id } if ingester_id == *follower_id) {
+                    return PingLeaderError::FollowerUnavailable;
+                }
+            }
+            PingLeaderError::LeaderUnavailable
+        })?;
+        Ok(())
+    }
+
+    /// Finds an available leader-follower pair to host a shard.
+    async fn find_leader(
```

Comment:
We might want to rename this `find_leader_and_follower` or `find_replicaset`.
I personally prefer `find_leader_and_follower`

Reply:
Roger.